### PR TITLE
PIMOB-1214: fix 3ds

### PR DESCRIPTION
**Pull request by harry-brown-cko**
## Proposed changes

- Use Checkout backend for 3DS URL logic
- Fix tests
- Merge 3.5.3 into 4.0.0
    - Including bringing webview events from 3.5.3 to 4.0.0

## Types of changes

What types of changes does your code introduce to Frames Ios?
_Put an `x` in the boxes that apply_

* [X] Bugfix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

* [X] Lint and unit tests pass locally with my changes
* [X] I have added tests that prove my fix is effective or that my feature works
* [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: Source/Core/Logging/FramesLogEvent.swift [hunk 1]
```diff
@@ -14,10 +15,32 @@ enum FramesLogEvent: Equatable {
         case serverError
         case tokenID
         case tokenType
+        case locale
+        case theme
+        case primaryBackgroundColor
```
Comment (deepesh-vasthimal-cko) [inline Source/Core/Logging/FramesLogEvent.swift:20]: I am expecting these will be thrown away as with the new UI release 4.0.0 these will be deleted anyways.

File: docs/Classes.html [hunk 2]
```diff
@@ -21,7 +21,7 @@
     <header class="header">
       <p class="header-col header-col--primary">
         <a class="header-link" href="index.html">
-          Frames 3.5.2 Docs
+          Frames 3.5.3 Docs
```
Comment (deepesh-vasthimal-cko) [inline docs/Classes.html:24]: Again, these doc changes will be overridden with what we already have in checkout repo.

**Review by deepesh-vasthimal-cko — APPROVED** [review]
Few minor comments

File: Checkout/Checkout/Source/3DSHandling/ThreeDSWKNavigationHelperDelegate.swift [hunk 3]
```diff
@@ -5,8 +5,9 @@
 //  Created by Daven.Gomes on 02/02/2022.
 //
 
-import Foundation
+import WebKit
 
 public protocol ThreeDSWKNavigationHelperDelegate: AnyObject {
+  func loaded(navigation: WKNavigation, success: Bool)
```
Comment (ehab-al-cko) [inline Checkout/Checkout/Source/3DSHandling/ThreeDSWKNavigationHelperDelegate.swift:11]: does `loaded` mean `didStartLoading` or `didFinishLoading` ?

File: Checkout/Checkout/Test/Stubs/StubThreeDSWKNavigationHelperDelegate.swift [hunk 4]
```diff
@@ -6,11 +6,18 @@
...
+import WebKit
 
 final class StubThreeDSWKNavigationHelperDelegate: ThreeDSWKNavigationHelperDelegate {
   private(set) var didReceiveResultCalledWith: [Result<String, ThreeDSError>] = []
 
   func threeDSWKNavigationHelperDelegate(didReceiveResult: Result<String, ThreeDSError>) {
     didReceiveResultCalledWith.append(didReceiveResult)
   }
+
+  private(set) var loadedCalledWith: (navigation: WKNavigation, success: Bool)?
```
Comment (ehab-al-cko) [inline Checkout/Checkout/Test/Stubs/StubThreeDSWKNavigationHelperDelegate.swift:18]: please group variables together

File: Checkout/Sample/Source/3DS/WebViewController.swift [hunk 5]
```diff
@@ -83,6 +83,8 @@ final class WebViewController: UIViewController {
 }
 
 extension WebViewController: ThreeDSWKNavigationHelperDelegate {
+  func loaded(navigation: WKNavigation, success: Bool) { }
```
Comment (ehab-al-cko) [inline Checkout/Sample/Source/3DS/WebViewController.swift:86]: there's no implementation, please add a comment for it

File: Source/Core/Logging/FramesEventLogger.swift [hunk 6]
```diff
@@ -37,6 +37,21 @@ final class FramesEventLogger: FramesEventLogging {
...
+                                             appPackageVersion: String,
+                                             uiDevice: UIDevice) -> RemoteProcessorMetadata {
+
+            return RemoteProcessorMetadata(productIdentifier: Constants.productName,
+                                           productVersion: Constants.version,
+                                           environment: environment.rawValue,
+                                           appPackageName: appPackageName,
+                                           appPackageVersion: appPackageVersion,
+                                           deviceName: uiDevice.modelName,
+                                           platform: "iOS",
```
Comment (ehab-al-cko) [inline Source/Core/Logging/FramesEventLogger.swift:51]:
```suggestion
                                           platform:  platform,
```

File: Source/Core/Logging/Theme.swift [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+//  Theme.swift
+//  Frames
+//
+//  Created by Harry Brown on 23/03/2022.
+//  Copyright © 2022 Checkout. All rights reserved.
+//
+
+import CheckoutEventLoggerKit
+
+class Theme: PropertyProviding, Equatable {
```
Comment (ehab-al-cko) [inline Source/Core/Logging/Theme.swift:11]:
```suggestion
final class Theme: PropertyProviding, Equatable {
```

File: Source/Core/ThreeDSWKNavigationHelperFactory.swift [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+//
+
+import Foundation
+import Checkout
+
+protocol ThreeDSWKNavigationHelperFactoryProtocol {
+    func build(successURL: URL?, failureURL: URL?) -> ThreeDSWKNavigationHelping
+}
+
+class ThreeDSWKNavigationHelperFactory: ThreeDSWKNavigationHelperFactoryProtocol {
```
Comment (ehab-al-cko) [inline Source/Core/ThreeDSWKNavigationHelperFactory.swift:16]:
```suggestion
final class ThreeDSWKNavigationHelperFactory: ThreeDSWKNavigationHelperFactoryProtocol {
```

File: Source/Extensions/UIColor+PropertyProviding.swift [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+//  Created by Harry Brown on 23/03/2022.
+//  Copyright © 2022 Checkout. All rights reserved.
+//
+
+import UIKit
+import CheckoutEventLoggerKit
+
+extension UIColor: PropertyProviding {
+    var properties: [FramesLogEvent.Property: AnyCodable] {
+        var red: CGFloat = 0.0, green: CGFloat = 0.0, blue: CGFloat = 0.0, alpha: CGFloat = 0.0
```
Comment (ehab-al-cko) [inline Source/Extensions/UIColor+PropertyProviding.swift:14]: this is black color, 
```suggestion
        var black: CGFloat = 0.0, green: CGFloat = 0.0, blue: CGFloat = 0.0, alpha: CGFloat = 0.0
```

File: Tests/Core/Extensions/UIBarStyle+stringValueTests.swift [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+//  FramesTests
+//
+//  Created by Harry Brown on 31/03/2022.
+//  Copyright © 2022 Checkout. All rights reserved.
+//
+
+import XCTest
+@testable import Frames
+
+final class UIBarStyle_stringValueTests: XCTestCase {
```
Comment (ehab-al-cko) [inline Tests/Core/Extensions/UIBarStyle+stringValueTests.swift:12]:
```suggestion
final class UIBarStyleTests: XCTestCase {
```

File: Source/Core/Logging/FramesEventLogger.swift (hunk 6 above)
Comment (harry-brown-cko) [reply Source/Core/Logging/FramesEventLogger.swift:53]: sorry wht do you mean by this?

File: Source/Core/Logging/Theme.swift (hunk 7 above)
Comment (harry-brown-cko) [reply Source/Core/Logging/Theme.swift:11]: if we change this one then `StubTheme` can't extend from `Theme`. This in turns means that we need to make a custom equals function on `FramesLogEvent`.

Not sure which is preferable

File: Source/Core/Logging/FramesEventLogger.swift (hunk 6 above)
Comment (ehab-al-cko) [reply Source/Core/Logging/FramesEventLogger.swift:53]: I suggest to use native device OS parameter. we may use payment for other OSs like WatchOS .

Comment (ehab-al-cko) [reply Source/Core/Logging/FramesEventLogger.swift:53]: I suggest to use native device OS parameter. we may use payment for other OSs like WatchOS .

**Review by ehab-al-cko — APPROVED** [review]

**Pull request merged**